### PR TITLE
Project contracts job

### PR DESCRIPTION
**Pull request by will0x0909**

### CODE REVIEW

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            a_transaction
+            for a_transaction in transactions
+            if a_transaction.receipt and a_transaction.receipt.status == 1
+        ]
+        traces = self._data_buff[Trace.type()]
+
+        self.transaction_map = {tnx.hash: tnx for tnx in transactions}
+        filtered_trace_list = [
+            trace for trace in traces if trace.trace_type.startswith(self.trace_type_prefix) and trace.status == 1
+        ]
```
Comment (shanshuo0918) [inline indexer/modules/custom/project_contracts/export_project_contracts_job.py:55]: Sinc you have filtered trace list and transaction, you can just combine them into your dataclass (without project_id)

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+        res = []
+        for project_id, deployers in self.project_deployers.items():
```
Comment (shanshuo0918) [inline indexer/modules/custom/project_contracts/export_project_contracts_job.py:58]: Then you iterate each contract creation, check whether it's a direct create or contract create, add project_id if exist.

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            for deployer in deployers:
+                res.extend(self.direct_create_contracts(project_id, deployer, filtered_trace_list, transactions))
+                res.extend(self.contract_create_contracts(project_id, deployer, filtered_trace_list))
+        self._collect_items(ProjectContractD.type(), res)
```
Comment (shanshuo0918) [inline indexer/modules/custom/project_contracts/export_project_contracts_job.py:62]: In the end, only keep those with project id?

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        # merge new contracts into exists
+        for pc in res:
+            self.project_contracts[pc.project_id].add(pc.address)
+        self.transaction_map.clear()
```
Comment (shanshuo0918) [inline indexer/modules/custom/project_contracts/export_project_contracts_job.py:66]: This transaction map can be a local variable?

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+class ExportProjectContractsJob(ExtensionJob):
+    dependency_types = [Transaction, Trace]
+    output_types = [ProjectContractD]
+    able_to_reorg = False
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self._batch_work_executor = BatchWorkExecutor(
```
Comment (shanshuo0918) [inline indexer/modules/custom/project_contracts/export_project_contracts_job.py:25]: I think you don't need this?

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            kwargs["batch_size"],
+            kwargs["max_workers"],
+            job_name=self.__class__.__name__,
+        )
+
+        self._is_batch = kwargs["batch_size"] > 1
+        self.db_service = kwargs["config"].get("db_service")
+        self.chain_id = self._web3.eth.chain_id
+        self.trace_type_prefix = "create"
```
Comment (shanshuo0918) [inline indexer/modules/custom/project_contracts/export_project_contracts_job.py:34]: This better to be a constant?

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py (hunk 5 above)
Comment (will0x0909) [reply indexer/modules/custom/project_contracts/export_project_contracts_job.py:25]: done

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py (hunk 4 above)
Comment (will0x0909) [reply indexer/modules/custom/project_contracts/export_project_contracts_job.py:66]: done

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py (hunk 1 above)
Comment (will0x0909) [reply indexer/modules/custom/project_contracts/export_project_contracts_job.py:47]: done

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py (hunk 2 above)
Comment (will0x0909) [reply indexer/modules/custom/project_contracts/export_project_contracts_job.py]: done

File: indexer/modules/custom/project_contracts/export_project_contracts_job.py (hunk 3 above)
Comment (will0x0909) [reply indexer/modules/custom/project_contracts/export_project_contracts_job.py:76]: now, only keep those contracts related to project

**Review by xuzh2024 — APPROVED** [review]

**Pull request merged**